src/Symfony/Component/Form/Extension/Core/Type/PasswordHasherExtension.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\Extension\Core\Type;
+
+use Symfony\Component\Form\AbstractTypeExtension;
+use Symfony\Component\Form\Extension\Core\EventListener\PasswordHasherListener;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+/**
+ * @author Sébastien Alfaiate <s.alfaiate@webarea.fr>
+ */
+class PasswordHasherExtension extends AbstractTypeExtension
+{
+    private $passwordHasher;
+    private $propertyAccessor;
+
+    public function __construct(UserPasswordHasherInterface $passwordHasher = null, PropertyAccessorInterface $propertyAccessor = null)
+    {
+        $this->passwordHasher = $passwordHasher;
+        $this->propertyAccessor = $propertyAccessor ?? PropertyAccess::createPropertyAccessor();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder->addEventSubscriber(new PasswordHasherListener($this->passwordHasher, $this->propertyAccessor));
```

Comment:
What is the performance impact of adding this recursing listener on **all** forms, even when they don't rely on PasswordType at all and even less on the new option ? The design used in this PR seems wrong to me, as it breaks the encapsulation of types.

Reply:
The event is added to all forms but only the root form is concerned (`$form->isRoot()` condition in the listener):
https://github.com/symfony/symfony/pull/42807/files#diff-fd2fa2abcca2add3e709b18c862ca809369f50fe351e1b381761fe9249f32288R48-R52

I don't think it is possible to determine that we are on the root form from the builder.
[`FormTypeCsrfExtension`](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Form/Extension/Csrf/Type/FormTypeCsrfExtension.php#L58) & [`FormTypeValidatorExtension`](https://github.com/symfony/symfony/blob/5.4/src/Symfony/Component/Form/Extension/Validator/Type/FormTypeValidatorExtension.php#L45) are working in the same way. The event is added to all forms and the condition `$form->isRoot()` is in the listener to target the root form.

Reply:
but you still recurse in the whole form tree.

The CSRF extension applies on all forms, but there is 2 very important difference:
- it only performs work for the root form. It does not recurse in the whole form tree
- CSRF protection is a feature that actually makes useful work for all root forms (if the CSRF protection is not enabled in the options, the listener is not registered and so does not run).

Reply:
I see. I have an idea to prevent recursion on the whole form.
I will give a try on it.

Reply:
I push a new version of the listener.

1. When the event is triggered on a password form field, if it is eligible, it is stored in a static property
2. Loop on the static property to hash the passwords

Reply:
I think I can make it more clean. I got more inspiration during the night XD
I will try to do the changes today.

Reply:
I fixed this using an event listener in the `PasswordType`, and then, another event only for the root form that will not require to recurse the whole form tree.